app/services/course_search.rb

```diff
@@ -0,0 +1,50 @@
+class CourseSearch
+  def initialize(filter:, recruitment_cycle_year:)
+    @filter = filter || {}
+    @recruitment_cycle_year = recruitment_cycle_year
+  end
+
+  class << self
+    def call(**args)
+      new(args).call
+    end
+  end
+
+  def call
+    scope = Course
```

Comment:
Just wondering does scope need to be ordered to be more performant?

ie

```ruby
faster = course
      .with_recruitment_cycle(recruitment_cycle_year)
      .findable

# vs 

slower = course
      .findable
      .with_recruitment_cycle(recruitment_cycle_year)

```

Reply:
No, these should be functionally equivalent.

Don't forget these are `ActiveRecord::Relation` objects, which means that a) the accumulate the parameters and only execute them once the results are requested, and b) we can examine what the SQL looks like:

```sql
> Course.findable.with_recruitment_cycle('2020').to_sql
SELECT "course".* FROM "course"
       INNER JOIN "course_site" ON "course_site"."course_id" = "course"."id"
       INNER JOIN "provider" ON "provider"."id" = "course"."provider_id"
       INNER JOIN "recruitment_cycle" ON "recruitment_cycle"."id" = "provider"."recruitment_cycle_id"
       WHERE "course_site"."status" = 'R'
             AND "course_site"."publish" = 'Y'
             AND "recruitment_cycle"."year" = '2020'


> Course.with_recruitment_cycle('2020').findable.to_sql
SELECT "course".* FROM "course"
       INNER JOIN "provider" ON "provider"."id" = "course"."provider_id"
       INNER JOIN "recruitment_cycle" ON "recruitment_cycle"."id" = "provider"."recruitment_cycle_id"
       INNER JOIN "course_site" ON "course_site"."course_id" = "course"."id"
       WHERE "recruitment_cycle"."year" = '2020'
             AND "course_site"."status" = 'R'
             AND "course_site"."publish" = 'Y'

```